src/clj_icat_direct/queries.clj

```diff
@@ -692,23 +692,25 @@
        AND d.data_name = ?"
 
    :folder-listing
-   "WITH user_access AS (SELECT *
-                           FROM r_objt_access
-                           WHERE user_id IN (SELECT g.group_user_id
-                                               FROM r_user_main u
-                                                 JOIN r_user_group g ON g.user_id = u.user_id
-                                               WHERE u.user_name = ? AND u.zone_name = ?)),
-         parent      AS (SELECT * FROM r_coll_main WHERE coll_name = ?)
-    SELECT DISTINCT c.coll_name || '/' || d.data_name AS full_path
-      FROM r_data_main d JOIN r_coll_main c ON d.coll_id = c.coll_id
-      WHERE c.coll_name IN (SELECT coll_name FROM parent)
-        AND d.data_id IN (SELECT object_id FROM user_access)
-    UNION
-    SELECT coll_name AS full_path
-      FROM r_coll_main
-      WHERE parent_coll_name IN (SELECT coll_name FROM parent)
-        AND coll_id IN (SELECT object_id FROM user_access)
-        AND coll_type != 'linkPoint'"
+   "WITH objs AS (SELECT coll_id AS object_id, coll_name AS full_path
+                  FROM r_coll_main
+                  WHERE parent_coll_name = ?
+                    AND coll_type != 'linkPoint'
+                  UNION
+                  SELECT d.data_id, c.coll_name || '/' || d.data_name
+                  FROM r_data_main d
+                           JOIN r_coll_main c ON d.coll_id = c.coll_id
+                  WHERE c.coll_name = ?),
+         user_access AS (SELECT DISTINCT object_id
+                         FROM r_objt_access
+                         WHERE user_id IN (SELECT g.group_user_id
+                                           FROM r_user_main u
+                                                    JOIN r_user_group g ON g.user_id = u.user_id
+                                           WHERE u.user_name = ? AND u.zone_name = ?)
+                           AND object_id IN (SELECT object_id FROM objs))
+    SELECT full_path
+    FROM objs
+    WHERE object_id IN (SELECT object_id FROM user_access)"
```

Comment:
I also tried the suggestion of using a join on `r_objt_access` and not using any CTEs, but I didn't see any performance improvements with that query (just using timing tests in data-info's save metadata endpoint), and this query seemed easier/faster for me to get my head around.

Reply:
That's probably fine. The place where this could become in issue (in either direction) is that CTEs are optimization fences. Basically, `objs`, `user_access`, and the top-level query are all independently handled by the query planner. Without the CTEs, it considers the whole query at once. If it's basically right about the numbers regardless, no difference, but it can sometimes be wrong about how to put the joins together (which would make the CTEs do better) or wrong about the consequences of stuff within the CTES (which can sometimes make the non-CTE do better). If you're seeing no difference right now, we're probably fine.

Reply:
rereading this I thought of another difference that I should say for completeness, plus the caveat that I could be forgetting a few more: in a non-CTE version sometimes things can be faster if the query is able to pipeline things directly into each other rather than making the CTE and then doing a "CTE scan". Pipelined query execution comes up more importantly in cases where the output is being sorted and going through a `LIMIT` at the end, which obviously isn't the case here -- in that case there's a big performance gain if the `ORDER BY` matches an index, such that the execution can only fetch relevant pages of the index and emit them in already-sorted order, rather than scanning the whole index and then sorting afterwards. But it can still come up within the topic of CTEs-as-optimization-fences sometimes.